src/bin/nydusd/fs_cache.rs

```diff
@@ -743,6 +747,166 @@ impl FsCacheHandler {
+            let volume_key = &child.file_name().to_str().unwrap().to_string()[1..];
+            let (cookie_dir, cookie_name) = self.generate_cookie_path(&path, volume_key, &blob_id);
+            let cookie_path = cookie_dir.join(&cookie_name);
+            if cookie_path.is_file() {
+                match self.inuse(&cookie_dir, &cookie_name) {
+                    Err(e) => {
+                        warn!(
+                            "blob {} call inuse err {}, cull failed!",
+                            cookie_path.display(),
+                            e
+                        );
+                        res = false;
+                    }
+                    Ok(true) => {
+                        warn!("blob {} in use, cull failed!", cookie_path.display());
+                        res = false;
+                    }
+                    Ok(false) => {
+                        if let Err(e) = self.cull(&cookie_dir, &cookie_name) {
+                            warn!(
+                                "blob {} call cull err {}, cull failed!",
+                                cookie_path.display(),
+                                e
+                            );
+                            res = false;
+                        }
+                    }
+                }
+            }
+        }
+        env::set_current_dir(&cwd_old)?;
+        if res {
+            Ok(())
+        } else {
+            Err(eother!("cull blob failed"))
+        }
+    }
+
+    #[inline]
+    fn hash_32(&self, val: u32) -> u32 {
+        val * 0x61C88647
+    }
+
+    #[inline]
+    fn rol32(&self, word: u32, shift: i32) -> u32 {
+        word << (shift & 31) | (word >> ((-shift) & 31))
+    }
+
+    #[inline]
+    fn round_up_u32(&self, size: usize) -> usize {
+        (size + 3) / 4 * 4
+    }
+
+    //address from kernel fscache_hash()
+    fn fscache_hash(&self, salt: u32, data: &[u8]) -> u32 {
+        assert_eq!(data.len() % 4, 0);
+
+        let mut x = 0;
+        let mut y = salt;
+        let mut buf_le32: [u8; 4] = [0; 4];
+        let n = data.len() / 4;
+
+        for i in 0..n {
+            buf_le32.clone_from_slice(&data[i * 4..i * 4 + 4]);
+            let a = unsafe { std::mem::transmute::<[u8; 4], u32>(buf_le32) }.to_le();
+            x ^= a;
+            y ^= x;
+            x = self.rol32(x, 7);
+            x += y;
+            y = self.rol32(y, 20);
+            y *= 9;
+        }
+        self.hash_32(y ^ self.hash_32(x))
+    }
+
+    fn generate_cookie_path(
+        &self,
+        volume_path: &Path,
+        volume_key: &str,
+        cookie_key: &str,
+    ) -> (PathBuf, String) {
+        //calc volume hash
+        let mut volume_hash_key: Vec<u8> =
+            Vec::with_capacity(self.round_up_u32(volume_key.len() + 2));
+        volume_hash_key.push(volume_key.len() as u8);
+        volume_hash_key.append(&mut volume_key.as_bytes().to_vec());
+        while volume_hash_key.len() != volume_hash_key.capacity() {
```

Comment:
Seems it can be  simplified as `volume_hash_key.resize(volume_hash_key.capacity())`?

Reply:
Fscache_hash() need the data is a multiple of four bytes in size , so just padding 0 here as what the kernel does.

Reply:
Understood , thanks, will fix